### ISSUE TITLE
Refactor to use coroutines

### ISSUE THREAD

**Issue by dhleong** [dhleong/judo#54]
Using coroutines would simplify a lot of code—especially around handling input on a single "UI" thread in a blocking manner. 

I was thinking about doing it as part of the big rendering refactor in progress on the `dhleong/render-refactor` branch for #51, but that's got quite enough going on.

**Comment by dhleong** [comment]
After playing with this idea for a bit, I'm going to close for now. Some simple synchronization was added in 1931cc1a5eb71ba713069cbb5df5807bb894fe6c that should pretty much cover the main problem of reading from the network on a background thread.

The biggest issue here is that our current blocking method of reading keys is actually a cooperative multithreading in that things that *need* to access the main thread (via `IJudoCore.onMainThread`) will still happen on the main thread even if we're blocking on `readKey()`. This means that calling `input()` in scripting won't block network events from firing—and since writes to `Buffer`s from the network occur in a background thread, that won't be blocked either.

In order to really make coroutines worthwhile, we need to use them everywhere—but some places, like scripting's `input()`—need blocking semantics, and some initial experiments suggest that `runBlocking` doesn't really allow for the kind of cooperation that we need. 

I experimented with using the current `onMainThread` as a `CoroutineDispatcher` and that sort of works, but it doesn't currently have enough bandwidth to make dispatching the Buffer writes to it performant enough—and there were also a handful of strange coroutine bugs cropping up.

Another annoyance is that the Kotlin compiler apparently barfs on inlining suspend functions... sometimes. This isn't a dealbreaker necessarily, but I ran into it a bunch working on the coroutine branch, and I'm not ready to introduce that confusion into an open source project.

All in all, there's not enough win here to really justify the increased complexity.
